### PR TITLE
Add assertions about <input> value dirty state

### PR DESCRIPTION
**Pull request by sophiebits**
Since this is an observable behavior and is hard to think about, seems good to have tests for this.

The expected value included in each test is the behavior that existed prior to #26546.

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js [hunk 1]
```diff
@@ -794,13 +815,16 @@ describe('ReactDOMInput', () => {
...
+    expect(isValueDirty(node)).toBe(false);
+
     ReactDOM.render(
       <input type="text" value={0} onChange={emptyFunction} />,
       container,
     );
 
-    const node = container.firstChild;
     expect(node.value).toBe('0');
+    expect(isValueDirty(node)).toBe(true);
```
Comment (sophiebits) [inline packages/react-dom/src/__tests__/ReactDOMInput-test.js:827]: Interestingly, this sole test fails on main with disableInputAttributeSyncing off:

```
  ● ReactDOMInput › should properly transition from an empty value to 0

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      825 |
      826 |     expect(node.value).toBe('0');
    > 827 |     expect(isValueDirty(node)).toBe(true);
          |                                ^
      828 |
      829 |     if (disableInputAttributeSyncing) {
      830 |       expect(node.hasAttribute('value')).toBe(false);
```

(It passes with the flag on.) So it regressed recently. Possibly related to the bug @rickhanlonii was debugging.

**Comment by BrendanWu** [comment]
Testing 1, 2, 3

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js (hunk 1 above)
Comment (BrendanWu) [reply packages/react-dom/src/__tests__/ReactDOMInput-test.js:827]: expect head

**Comment by sophiebits** [comment]
@BrendanWu please don't use issues in popular projects for your own testing – thanks!

**Comment by BrendanWu** [comment]
Working on it

**Review by sebmarkbage — APPROVED** [review]

File: packages/react-dom/src/__tests__/ReactDOMInput-test.js [hunk 2]
```diff
@@ -26,6 +26,16 @@ describe('ReactDOMInput', () => {
     node.dispatchEvent(new Event(type, {bubbles: true, cancelable: true}));
   }
 
+  function isValueDirty(node) {
+    // Return the "dirty value flag" as defined in the HTML spec. Cast to text
+    // input to sidestep complicated value sanitization behaviors.
+    const copy = node.cloneNode();
```
Comment (sebmarkbage) [inline packages/react-dom/src/__tests__/ReactDOMInput-test.js:32]: Interesting that cloning clones the dirty flag.

**Comment by sebmarkbage** [comment]
I'm very excited about this. Does something like this make sense on Textarea too?

**Comment by sophiebits** [comment]
Yes, and dirty checked for checkbox/radio.

**Pull request merged**